design/0003-byop.md

```diff
@@ -0,0 +1,198 @@
+|---------|------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| true    | ""               | Install linkerd-prometheus, and configure control-plane components to use it.                                                                                                                                                                    |
+| false   | "prometheus.xyz" | No linkerd-prometheus, control-plane components will be configured to url                                                                                                                                                                        |
+| true    | "prometheus.xyz" | Install linkerd-prometheus, but stil configure control plane to use different url, useful when linkerd-prometheus is configured to remote-write into long term storage solution at url, Thus  powering dashboard and CLI through long term storage. |
+| false   | ""               | No linkerd-prometheus, and no URL to power dashboards. (Should we even allow this?)                                                                                                                                                              |
+
+#### User Expereince
+
+Currently, `public-api`, `grafana` and `heartbeat` are the only three
+components which are configured with a `prometheus-url` parameter,
+All the Dashboards and CLI are powered through the `public-api`.
+This allows us to swap prometheus without breaking any exisiting linkerd
+tooling. Though it is expected that the `scrape-config` is configured
+correctly, when this path is choosen.
+
+If `prometheus-url` is not accessible, components like `public-api` should not fail
+but just log some warning. So, that users can still access things like TAP, Checks, etc through the dashboard and CLI.
+
+#### Linkerd CLI Tooling
+
+When users get their own Prometheus, It would be really great to have
+some tooling that allows users to check and troubleshoot if there are any
+problems.
+
+*linkerd check*
+
+Checks can be added to report if the prometheus is configured correctly and is replying to queries.
+
+- Check if the configured prometheus instance is healthy.
+- if the relevant metrics are present in prometheus by querying, and warn the users about the same. User's
+don't have to supply the `prometheusUrl` as we can access the same from `linkerd-config-addons` Configmap.
+- More checks can be added here, if there are some known
+configurations that don't go well with linkerd.
+
+*linkerd prom-config*
+
+A additional command can be provided in the CLI, which outputs the prometheus-config corresponding to that version of linkerd, which would
+have the relevant scrape-configs and default configuration.
+
+
+### Prior art
```

Comment:
It would be helpful to describe alternative solutions that we considered and explored and why they were not suitable.  

For examples:
* agressive tuning of linkerd-prometheus
* exposing more tuning knobs of linkerd-prometheus
* dropping labels

Reply:
> exposing more tuning knobs of linkerd-prometheus

This rfc address this by exposing the `global` field at `prometheus.global` where most tuning around retention period, etc can be configured. 

All the tuning stuff though being very helpful to make the linkerd-prometheus more stable, managable and not run into situations like OOM, etc when there is sudden spikes in queries, etc still won't address the user's question of getting their own prometheus, which we definitely heard a lot! (definitely not suggestible though)

I added this suggestion. Feel free to comment, if you would like me to add anything!